### ISSUE TITLE
You must feed a value for placeholder tensor 'output' with dtype float and shape [?,1]

### ISSUE THREAD

**Issue by GertjanBrouwer** [serizba/cppflow#30]
I have a neural net that has an input of [None, 353] and an ouput of [None, 1]. This however throws this error: `You must feed a value for placeholder tensor 'output' with dtype float and shape [?,1]` The output tensor in cppflow has shape `-1, 1` and the input has `-1, 353`.
What is a proper way to solve this?

I have tried saving the model.pb again with the shape [1, 353] and [1, 1] but the same error came up.

**Comment by serizba** [comment]
Hi,

Could you provide a small example so it is possible to repeat and reproduce your problem? I cannot get the source of the problem without having a look at the Python code to generate the .pb and the C++ code to perform the inference.

If possible, try to post a small piece of code with just the minimum code to be able to reproduce the error.

**Comment by GertjanBrouwer** [comment]
Hey thanks for the quick respone, I really appreciate that. 
I have created a Google Colab notebook (just like python notebook) in which I create the model, train and save etc. Here is a link of a copy of that notebook: [link](https://colab.research.google.com/drive/1teysv0FItq5BQlxBq4kNjA2-u6GbEHIP?usp=sharing). You can edit it also if you want. I can also provide some example data if you like, but just random values should work as training. 

The C++ is code is just this:
```C++
        Model model("./model/model.pb");
	model.restore("./model/model.ckpt");
	Tensor input(model, "input");
	Tensor prediction(model, "output");
	
        Board* board = new Board();
	board->setBoard("4b3/8/8/8/8/8/2B3P1/8 b - - 0 1");
	Feature* feature = new Feature(board);
	std::vector<double> featureVector = feature->generate();
	std::vector<float> featureVectorFloat(featureVector.begin(), featureVector.end());
	
        input.set_data(featureVectorFloat);
	model.run(input, prediction);
	auto result = prediction.get_data<float>();
```

Then at the model.cpp on this line: `TF_SessionRun(this->session, nullptr, io.data(), iv.data(), inputs.size(), oo.data(), ov, outputs.size(), nullptr, 0, nullptr, this->status);` it fails with the error in the title.

**Comment by serizba** [comment]
Hey, I think the problem is not related with the shapes. I would say that when you declare your model in the Python code you are calling 'output' to the actual target of the model (the labels or y_hat).

In your code `self._y` is called 'output' but actually it is not the output, is the data you feed to compare with. The actual output is `self.y`, the result of the network. 

I would try to change the name of `self._y` to something else, like 'target' and rename `self.y` to 'output`. You can use [tf.identity](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/identity). Like this:

```python
# Input variable and label
self.x = tf.placeholder(tf.float32, shape=[None, 353], name="input")
self._y = tf.placeholder(tf.float32, shape=[None, 1], name="target")

# Output
self.y = tf.identity(self.buildNetwork(self.x), name='output')
```

Now the C++ code should work. When you ask for the 'output' tensor, you will get your desired tensor, the result of applying the network.

Hope it works.

**Comment by GertjanBrouwer** [comment]
Thanks for the fast response, it still has the same error after building and training the network like that. Any other suggestions?

**Comment by serizba** [comment]
I have tried your code and I got the same error. I have tried to change what I suggested and now it is working. I guess you or I are missing something. I did not train, just create the model, initialize it and run it on C++ and it is working. I show you here the code I have used.

### Python
I just changed the name of self._y to 'target' and the name of self.y to 'output'. I also changed the path where the model is saved.
```python
tf.disable_eager_execution()
class Model(object):

    def __init__(self, restore=None):

        self.index = 0
        # Input variable and label
        self.x = tf.placeholder(tf.float32, shape=[None, 353], name="input")
        self._y = tf.placeholder(tf.float32, shape=[None, 1], name="target")

        # Output
        self.y = tf.identity(self.buildNetwork(self.x), name='output')

        # Build the optimizer
        self.cost = tf.reduce_mean(tf.abs(tf.subtract(self.y, self._y)))
        self.optimizer = tf.train.AdamOptimizer().minimize(self.cost)

        # Add ops to save and restore all the variables.
        self.saver = tf.train.Saver()
        # Operation to initialize the variables
        self.init = tf.global_variables_initializer()  
        # Start session
        self.sess = tf.Session()

        self.sess.run(self.init)

        #self.sess.run(self.y, feed_dict={self.x: feature})
        with open('model.pb', 'wb') as f:
            f.write(tf.get_default_graph().as_graph_def().SerializeToString())



    # Build the network structure
    def buildNetwork(self, x):

        # Weights and biases for the first hidden layer
        globalW = tf.Variable(tf.random_normal([17, 11], stddev=0.35),
                              name="globalW")
        globalB = tf.Variable(tf.random_normal([11], stddev=0.35),
                               name="globalB")
        pieceW = tf.Variable(tf.random_normal([208, 100], stddev=0.35),
                              name="pieceW")
        pieceB = tf.Variable(tf.random_normal([100], stddev=0.35),
                               name="pieceB")
        squareW = tf.Variable(tf.random_normal([128, 60], stddev=0.35),
                               name="squareW")
        squareB = tf.Variable(tf.random_normal([60], stddev=0.35),
                               name="squareB")

        # Weights and biases for the second hidden layer
        w2 = tf.Variable(tf.random_normal([171, 75], stddev=0.35),
                              name="W2")
        b2 = tf.Variable(tf.random_normal([75], stddev=0.35),
                              name="B2")

        # Weights and biases for the output layer
        w3 = tf.Variable(tf.random_normal([75, 1], stddev=0.35),
                         name="W3")
        b3 = tf.Variable(tf.random_normal([1], stddev=0.35),
                         name="B3")


        # Network operation

        # First hidden layer
        globalOutput = tf.add(tf.matmul(tf.slice(x, [0, 0], [-1, 17]), globalW), globalB)
        print(globalOutput.shape)
        pieceOutput = tf.add(tf.matmul(tf.slice(x, [0, 17], [-1, 208]), pieceW), pieceB)
        squareOutput = tf.add(tf.matmul(tf.slice(x, [0, 17+208], [-1, -1]), squareW), squareB)
        layer1Output = tf.concat([globalOutput, pieceOutput, squareOutput], axis=1)

        # Second hidden layer
        layer2Input = tf.nn.relu6(layer1Output)
        layer2Output = tf.add(tf.matmul(layer2Input, w2), b2)

        # Output layer
        layer3Input = tf.nn.relu6(layer2Output)
        layer3Output = tf.add(tf.matmul(layer3Input, w3), b3)
        return tf.tanh(layer3Output)


model = Model()

```

### C++
Just a small piece of code (based on examples/load_model and on what you provided in your message) to load the model and run with some data.
```c++
#include "../../include/Model.h"
#include "../../include/Tensor.h"

#include <numeric>
#include <iomanip>

int main() {
    Model model("../model.pb");
    model.init();


    Tensor input(model, "input");
    Tensor prediction(model, "output");
    
    std::vector<float> data(353);
    std::iota(data.begin(), data.end(), 0);
    input.set_data(data);

    model.run(input, prediction);
    auto result = prediction.get_data<float>();

    std::cout << "END" << std::endl;

}
```

Could you try if this piece of code is working for you?

**Comment by GertjanBrouwer** [comment]
I forgot to name the Tensor 'outputy', after that I get this error: `You must feed a value for placeholder tensor 'input_2' with dtype float and shape [?,353]` I actually named the input placeholder `input`, so I don't know why it is doing this. I will try your example now.

**Comment by serizba** [comment]
Does the error appear executing the python script or the c++ code?

Which version of TensorFlow and on which OS are you running?

**Comment by GertjanBrouwer** [comment]
It appears when running C++. I added this: `std::cout << TF_Message(status);` to the status_check to get that output. Windows 10 Pro build 18362.900, Tensorflow from the download link you provided with cpu-only support version 1.15

**Comment by serizba** [comment]
When i show all variables in the Python snippet that I posted I don't get an operation called 'input_2'. There is one 'input', one 'target' and several other operations but there isn't a 'input_2'. May you be trying a different code snippet?

**Comment by GertjanBrouwer** [comment]
That is weird, when I copy-paste your snippet it does not have an input_2. Where did you download the Tensorflow from that you include in model.h and tensor.h? I have a tensorflow folder in my project which I reference, possibly that might have to do something with it?

**Comment by serizba** [comment]
From the official [TensorFlow C API website](https://www.tensorflow.org/install/lang_c). Then, does the snippet that I posted work? I think the error is not related with TF versions but rather with small typos or bugs in the code (like naming wrong a variable).

**Comment by GertjanBrouwer** [comment]
The problem with this on Windows is that you can't install the same you would on Linux and Mac. So I import the official C .lib file in my project and add the .dll to my PATH variable. But the include in my code still have to be changed to a header file, which is not installed globally. Thats why I also have to download Tensorflow repo and include header files from there.

**Comment by serizba** [comment]
But I guess that's not related to the errors.

**Comment by GertjanBrouwer** [comment]
Ok I got yours to work somehow, you are probably right. I still don't know why tho. 
I copy pasted yours into my colab and that works, but what is the difference other than the names?

**Comment by serizba** [comment]
Names are how you declare which operation or tensor you want to use or feed. With the code that you first showed me, you were trying to get a placeholder as an output, while it was supposed to be an input, the target. I changed the name of the _y placeholder to 'target' and y to 'output', so when you ask for 'output' in the c++ code, it will run the operation related to y, that is, the whole network. Before you were trying to run the _y operation, but that's not an operation, is just a placeholder, so you cannot get anything from it.

**Comment by GertjanBrouwer** [comment]
Alright, that actually seems to be the problem. However, when doing inference it always returns -1.000000 instead of a value between -1 and 1. I did do inference in python on a single sample which did work. Any suggestions?

**Comment by serizba** [comment]
Did you save and restore the checkpoint properly?

**Comment by GertjanBrouwer** [comment]
Yeh i think so, i did this:
```C++
        Model model("./totalmodel/model.pb");
	//model.init();
	model.restore("./totalmodel/model.ckpt");
	Tensor input(model, "inputoud");
	Tensor prediction(model, "outputoud");
	Board* board = new Board();

	board->setBoard("r1b1k2r/2p2ppp/p1np1n2/1pb1p3/4P3/1BNP1N1P/1PP2PP1/R1B1R1K1 b kq - 0 1");

	Feature* feature = new Feature(board);
	std::vector<double> featureVector = feature->generate();
	std::vector<float> featureVectorFloat(featureVector.begin(), featureVector.end());
	input.set_data(featureVectorFloat);
	model.run(input, prediction);
	auto result = prediction.get_data<float>();
```

And i did this in python:
```python
self.saver.save(self.sess, "drive/My Drive/Winchess-data/newest-model/model.ckpt")
```

The saver is done when creating the model like so:
```python
self.saver = tf.train.Saver()
```

**Comment by GertjanBrouwer** [comment]
I have saved the checkpoint. When restoring it in Python it works, in C++ it only returns -1 or 1.
https://wetransfer.com/downloads/606f20cc885b4f422f6cf9aa09ea438920200611192224/09d37571e81abb79f68fcbc84eb49c2020200611192241/e24a56 here you have my saved files so you can try yourself

**Comment by serizba** [comment]
I tried your model in C++ and Python and using the same input I got the same output. For example feeding an input tensor of 353 values with the number (-0.05) I got an output of 0.51. I chose that number just to show an example with a value between -1 and 1. I show you the code I used:

### c++
```c++

#include "../../include/Model.h"
#include "../../include/Tensor.h"

#include <numeric>
#include <iomanip>

int main() {
    Model model("../newnorm.pb");
    model.restore("../newnorm.ckpt");
    //model.init();

    Tensor input(model, "inputoud");
    Tensor prediction(model, "outputoud");
    
    std::vector<float> data(353);
    for (int i=0; i<353; i++)
        data[i] = -0.05;
    input.set_data(data);

    model.run(input, prediction);
    std::cout << "Output: " << prediction.get_data<float>()[0] << std::endl;
}
```

### Python
```python
import tensorflow.compat.v1 as tf
import numpy as np
tf.disable_eager_execution()

sess=tf.Session()  
saver = tf.train.import_meta_graph('newnorm.ckpt.meta')
saver.restore(sess,tf.train.latest_checkpoint('./'))

graph = tf.get_default_graph()
input = graph.get_tensor_by_name("inputoud:0")
output = graph.get_tensor_by_name("outputoud:0")

output = sess.run(output, feed_dict={input: np.ones((1, 353)) * -0.05})

print(output)
```

May your problem be that you are not feeding the same values in your c++ code and your python code. If both of the above code snippets output the same for you that means that everything works fine. I would recommend you to review your code, and ensure that you are feeding the same values in Python and in C++.

Also, the topic of the issue has deviated from the initial error. I think is not related to the repository, but rather with the use of TensorFlow.

**Comment by GertjanBrouwer** [comment]
Alright I got it to work. I am doing the training in Google Colab with a link to my Google Drive. Apparently there was some issue with either saving with the same name or the TF session was not cleared properly. I got it working by resetting the runtime(colab specific) after each training session or reload of the model. 

Thanks a lot for your help

**Comment by serizba** [comment]
I'm glad you made it work. I proceed to close the issue.
